File: Leetcode/dfs_recursion_backtracking/216_CombinationSum_III/Solution.py
class Solution(object):
    def combinationSum3(self, k, n):
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """
        def dfs(start, total, temp):
            if total>n or len(temp)>k:
                return
            if total==n and len(temp)==k:
                res.append(temp)
                return
            for i in range(start, 10):
                dfs(i+1, total+i, temp+[i])
        res =[]
        dfs(1, 0, [])
        return res

File: Leetcode/dfs_recursion_backtracking/216_CombinationSum_III/Solution.java
class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>>  res = new LinkedList<>();
        dfs(res, new LinkedList<Integer>(), n, k, 1, 0);
        return res;
    }
    private void dfs(List<List<Integer>> res, LinkedList<Integer> temp,
                    int n, int k, int start, int total){
        if( total>n || temp.size()>k )
            return;
        if( total==n && temp.size()==k ){
            res.add(new LinkedList<Integer>(temp));
            return;
        }
        for(int i = start; i<10; ++i){
            temp.add(i);
            dfs(res, temp, n, k, i+1, total+i);
            temp.pollLast();
        }
    }
}
